src/cgroups/v2/systemd_manager.rs

```diff
@@ -0,0 +1,304 @@
+            }
+
+            // last component cannot have subtree_control enabled due to internal process constraint
+            // if this were set, writing to the cgroups.procs file will fail with Erno 16 (device or resource busy)
+            if components.peek().is_some() {
+                Self::write_controllers(&current_path, &controllers)?;
+            }
+        }
+
+        write_cgroup_file(full_path.join(CGROUP_PROCS), &pid.to_string())?;
+        Ok(full_path)
+    }
+
+    fn get_available_controllers<P: AsRef<Path>>(
+        &self,
+        cgroups_path: P,
+    ) -> Result<Vec<ControllerType>> {
+        let controllers_path = self.root_path.join(cgroups_path).join(CGROUP_CONTROLLERS);
+        if !controllers_path.exists() {
+            bail!(
+                "cannot get available controllers. {:?} does not exist",
+                controllers_path
+            )
+        }
+
+        let mut controllers = Vec::new();
+        for controller in fs::read_to_string(&controllers_path)?.split_whitespace() {
+            match controller {
+                "cpu" => controllers.push(ControllerType::Cpu),
+                "io" => controllers.push(ControllerType::Io),
+                "memory" => controllers.push(ControllerType::Memory),
+                "pids" => controllers.push(ControllerType::Pids),
+                _ => continue,
+            }
+        }
+
+        Ok(controllers)
+    }
+
+    fn write_controllers(path: &Path, controllers: &Vec<String>) -> Result<()> {
+        for controller in controllers {
+            common::write_cgroup_file_str(path.join(CGROUP_SUBTREE_CONTROL), controller)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl CgroupManager for SystemDCGroupManager {
+    fn apply(&self, linux_resources: &LinuxResources, pid: Pid) -> Result<()> {
+        // Dont attach any pid to the cgroup if -1 is specified as a pid
+        if pid.as_raw() == -1 {
+            return Ok(());
+        }
+        let full_cgroup_path = self.create_unified_cgroup(pid)?;
+
+        for controller in CONTROLLER_TYPES {
+            match controller {
+                ControllerType::Cpu => Cpu::apply(linux_resources, &full_cgroup_path)?,
+                ControllerType::CpuSet => CpuSet::apply(linux_resources, &full_cgroup_path)?,
+                ControllerType::HugeTlb => HugeTlb::apply(linux_resources, &&full_cgroup_path)?,
+                ControllerType::Io => Io::apply(linux_resources, &&full_cgroup_path)?,
+                ControllerType::Memory => Memory::apply(linux_resources, &full_cgroup_path)?,
+                ControllerType::Pids => Pids::apply(linux_resources, &&full_cgroup_path)?,
+            }
+        }
+
+        Ok(())
+    }
+
+    fn remove(&self) -> Result<()> {
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn expand_slice_works() -> Result<()> {
+        let manager = SystemDCGroupManager::new(
+            Path::new("/sys/fs/cgroup").to_path_buf(),
```

Comment:
You can use PathBuf::from("...") to create a PathBuf directly

Reply:
Done.